From 35bd9206dc3376cf8c531d6e5c3d7255bf1dcb6a Mon Sep 17 00:00:00 2001
From: Antonio Rivera <antoniori@google.com>
Date: Thu, 02 May 2024 22:28:28 +0000
Subject: [PATCH] Fix a cast linux crash in UrlRequestRulesReceiver.

https://chromium-review.googlesource.com/c/chromium/src/+/5407368/4/chromecast/cast_core/runtime/browser/runtime_application_service_impl.cc#b323
removed the line that sets params->enable_url_rewrite_rules to false in
RuntimeApplicationServiceImpl::CreateCastWebView(). That line was
specifically added to prevent this crash, in crrev.com/c/4052724.

The crash occurs because, if enable_url_rewrite_rules is true, both
CastWebContentsImpl and cast_receiver::ApplicationControlsImpl attempt
to create UrlRequestRewriteRulesManager instances. When a second
instance tries to connect to UrlRequestRulesReceiver, it crashes since a
mojo connection already exists.

See the bug below for the full stack trace.

Bug: b:336395582
Change-Id: I04b8ac0b5bdee45c5d54bb988133dfa8c8c15869
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5479989
Reviewed-by: Simeon Anfinrud <sanfin@chromium.org>
Reviewed-by: Shawn Quereshi <shawnq@google.com>
Commit-Queue: Antonio Rivera <antoniori@google.com>
Cr-Commit-Position: refs/heads/main@{#1295798}
---

--- a/chromecast/cast_core/runtime/browser/runtime_application_service_impl.cc
+++ b/chromecast/cast_core/runtime/browser/runtime_application_service_impl.cc
@@ -9,6 +9,8 @@
 #include "base/strings/stringprintf.h"
 #include "base/task/bind_post_task.h"
 #include "base/task/sequenced_task_runner.h"
+#include "build/build_config.h"
+#include "build/chromecast_buildflags.h"
 #include "chromecast/base/metrics/cast_metrics_helper.h"
 #include "chromecast/browser/cast_web_service.h"
 #include "chromecast/browser/cast_web_view.h"
@@ -333,6 +335,13 @@
       GetFlagEntry(feature::kCastCoreIsRemoteControlMode,
                    config_.extra_features(), /*default_value=*/false);
   params->enabled_for_dev = IsEnabledForDev();
+#if BUILDFLAG(ENABLE_CAST_RECEIVER) && BUILDFLAG(IS_LINUX)
+  // cast_receiver::ApplicationControlsImpl constructs an instance of
+  // url_rewrite::UrlRequestRewriteRulesManager. CastWebContentsImpl should NOT
+  // construct its own instance, or UrlRequestRulesReceiver will crash when a
+  // second mojo connection is attempted.
+  params->enable_url_rewrite_rules = false;
+#endif  // BUILDFLAG(ENABLE_CAST_RECEIVER) && BUILDFLAG(IS_LINUX)
   params->enable_touch_input = IsTouchInputAllowed();
   params->log_js_console_messages =
       GetFlagEntry(feature::kCastCoreLogJsConsoleMessages,
